### PR TITLE
Application domain implementation & single repository pattern

### PR DESCRIPTION
**Pull request by AhmadKabakibi**
#### What does this PR do?
- [x] Add EventStore to application layer
- [x] Add user use cases
- [x] Add Domain schema structure with properties validation 
- [x] Add user repository to infrastructure layer
- [x] Add API Route GET /api/users
- [x] Add API Route POST /api/users
- [x] Add API Route PUT /api/users
- [x] Add API Route DEL /api/users

#### Any background context you want to provide?
 I’m trying to cover here:
- separate core business logic from implementation details

### CODE REVIEW

File: src/app/EventStore.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+const define = Object.defineProperty
+
+class EventStore extends EventEmitter {
+  static setOutputs(outputs) {
+    define(this.prototype, 'outputs', {
+      value: createOutputs(outputs)
+    })
+  }
+
+  on(output, handler) {
```
Comment (dasblitz) [inline src/app/EventStore.js]: Maybe add on off() as well to cleanup listeners, or is there no use-case for this in NodeJS land and am I just being too DOM? (Dutch pun intended :) )

File: src/app/EventStore.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import EventEmitter from 'events'
+const define = Object.defineProperty
+
+class EventStore extends EventEmitter {
+  static setOutputs(outputs) {
+    define(this.prototype, 'outputs', {
```
Comment (dasblitz) [inline src/app/EventStore.js]: Nice dont think I've seen defineProperty being used in the wild before, so this basically creates an immutable `outputs` prop on the EventStore? 

I'm a bit confused as to where this `outputs` 'lives' as my OOP knowledge is lacking:

Once EventStore.setOutputs() is called will it be available statically as `EventStore.outputs` or do I need to instantiate a `const eventStore = new EventStore()` first and can I then access it as `eventStore.outputs`?

Comment (grumpyoldman-io) [reply src/app/EventStore.js]: EventStore will get a function called `setOutputs` which will create (on calling that function) the EventStore.outputs, before that there is no `.outputs` on the EventStore

Comment (grumpyoldman-io) [reply src/app/EventStore.js]: Oeps, I deleted my msg, but I agree with you, this feels a bit like the long way around. @AhmadKabakibi  lets look at it.

Comment (dasblitz) [reply src/app/EventStore.js]: Sounds good to me, I'll be back in the office on Monday. I'm not against OOP btw, just don't know too much about it. I thought using `Object.defineProperty` alone would be enough to not make it writable from class instances. Good topics for learning/talks I guess 👍

**Comment by grumpyoldman-io** [comment]
just needs some specs, afterwards lets see if we can squash / clean up the commit history. There is a lot of duplicate commits.

**Comment by grumpyoldman-io** [comment]
just needs some specs, afterwards lets see if we can squash / clean up the commit history. There is a lot of duplicate commits.

**Comment by grumpyoldman-io** [comment]
just needs some specs, afterwards lets see if we can squash / clean up the commit history. There is a lot of duplicate commits.

**Comment by AhmadKabakibi** [comment]
I can comment here :D

**Pull request closed without merging**